benchmark/tools/compare.py

```diff
@@ -0,0 +1,261 @@
+#!/usr/bin/env python3
+import sys
+import json
+import argparse
+import math
+import pandas as pd
+
+parser = argparse.ArgumentParser(description="Compare to Ginkgo benchmark outputs")
+parser.add_argument(
+    "--outliers", action="store_true", help="List outliers from the results"
+)
+parser.add_argument(
+    "--outlier-threshold",
+    type=float,
+    default=10,
+    help="At what percentage of deviation (above or below) should outliers be reported",
+)
+parser.add_argument(
+    "--outlier-count",
+    type=int,
+    default=1000,
+    help="How many outliers should be reported per benchmark",
+)
+parser.add_argument("--output", choices=["json", "csv", "markdown"], default="json")
+parser.add_argument("baseline")
+parser.add_argument("comparison")
+args = parser.parse_args()
+keys = {"stencil", "size", "filename", "n", "r", "k", "m"}
+comparison_keys = {"time", "storage", "iterations"}
+suffix = ".ratio"
+
+
+def key_to_json(key: tuple) -> dict:
+    """Restore a JSON output from a key tuple"""
+    result = {}
+    for key_name, key_val in zip(keys, key):
+        if key_val is not None:
+            result[key_name] = key_val
+    return result
+
+
+def key_to_str(key: tuple) -> str:
+    """Restore a JSON output string from a key tuple"""
+    mapped = key_to_json(key)
+    if "filename" in mapped.keys():
+        return mapped["filename"]
+    if "stencil" in mapped.keys():
+        return "stencil({}, {})".format(mapped["stencil"], mapped["size"])
+    return json.dumps(mapped).replace('"', "")
+
+
+def parse_json_matrix(filename: str) -> dict:
+    """Parse a JSON file into a key -> test_case dict"""
+    parsed = json.load(open(filename))
+    result = {}
+    assert isinstance(parsed, list)
+    for case in parsed:
+        assert isinstance(case, dict)
+        assert not keys.isdisjoint(case.keys())
+        dict_key = tuple(case.get(key, None) for key in keys)
+        if dict_key in result.keys():
+            print(
+                "WARNING: Duplicate key {}".format(key_to_str(dict_key)),
+                file=sys.stderr,
+            )
+        result[dict_key] = case
+    return result
+
+
+def warn_on_inconsistent_keys(baseline: dict, comparison: dict, context: str):
+    """Print a warning message for non-matching keys between baseline/comparison using the given context string"""
+    baseline_only = set(baseline.keys()).difference(comparison.keys())
+    comparison_only = set(comparison.keys()).difference(baseline.keys())
+    for dict_key in baseline_only:
+        print(
+            "WARNING: Key {} found in baseline only in context {}".format(
```

Comment:
Instead of crafting warning messages manually and printing it out to sys.stderr, you could use pythons logging module https://docs.python.org/3/library/logging.html (or check warnings  https://docs.python.org/3/library/warnings.html together)

Reply:
I'd like to test everything using capsys and pytest, is it easy to capture warnings?